src/bun.js/bindings/NodeVM.cpp

```diff
@@ -741,8 +764,50 @@ constructScript(JSGlobalObject* globalObject, CallFrame* callFrame, JSValue newT
         JSC::StringSourceProvider::create(sourceString, JSC::SourceOrigin(WTF::URL::fileURLWithFileSystemPath(options.filename)), options.filename, JSC::SourceTaintedOrigin::Untainted, TextPosition(options.lineOffset, options.columnOffset)),
         options.lineOffset.zeroBasedInt(), options.columnOffset.zeroBasedInt());
     RETURN_IF_EXCEPTION(scope, {});
-    NodeVMScript* script = NodeVMScript::create(vm, globalObject, structure, source);
-    return JSValue::encode(JSValue(script));
+
+    const bool produceCachedData = options.produceCachedData;
+    auto cachedData = WTFMove(options.cachedData);
+    auto filename = options.filename;
+
+    NodeVMScript* script = NodeVMScript::create(vm, globalObject, structure, source, WTFMove(options));
+
+    if (!cachedData.empty()) {
+        JSC::ProgramExecutable* executable = script->cachedExecutable();
+        if (!executable) {
+            executable = script->createExecutable();
+        }
+        ASSERT(executable);
+
+        JSC::LexicallyScopedFeatures lexicallyScopedFeatures = globalObject->globalScopeExtension() ? JSC::TaintedByWithScopeLexicallyScopedFeature : JSC::NoLexicallyScopedFeatures;
+        JSC::SourceCodeKey key(source, {}, JSC::SourceCodeType::ProgramType, lexicallyScopedFeatures, JSC::JSParserScriptMode::Classic, JSC::DerivedContextType::None, JSC::EvalContextType::None, false, {}, std::nullopt);
+        Ref<JSC::CachedBytecode> cachedBytecode = JSC::CachedBytecode::create(cachedData, nullptr, {});
+        JSC::UnlinkedProgramCodeBlock* unlinkedBlock = JSC::decodeCodeBlock<UnlinkedProgramCodeBlock>(vm, key, WTFMove(cachedBytecode));
+
+        if (!unlinkedBlock) {
+            script->cachedDataRejected(TriState::True);
+        } else {
+            JSC::JSScope* jsScope = globalObject->globalScope();
+            JSC::CodeBlock* codeBlock = nullptr;
+            {
+                // JSC::ProgramCodeBlock::create() requires GC to be deferred.
+                DeferGC deferGC(vm);
+                codeBlock = JSC::ProgramCodeBlock::create(vm, executable, unlinkedBlock, jsScope);
+            }
+            JSC::CompilationResult compilationResult = JIT::compileSync(vm, codeBlock, JITCompilationEffort::JITCompilationCanFail);
+            if (compilationResult != JSC::CompilationResult::CompilationFailed) {
+                executable->installCode(codeBlock);
+                script->cachedDataRejected(TriState::False);
+            } else {
+                script->cachedDataRejected(TriState::True);
+            }
+        }
+    } else if (produceCachedData) {
+        script->cacheBytecode();
+        // TODO(@heimskr): is there ever a case where bytecode production fails?
```

Comment:
https://github.com/WebKit/WebKit/pull/44754

although i don't know if it applies to this situation